File: frontend/bridge/pcui/capabilities.py
"""Capabilities handshake assembly (frontend/design.md §3): one pass at startup over the four
skills' `schema` (+ team `vocab`) output AND their shipped data trees — a fingerprint must
identify the actually-installed contract *and data version* (a meta refresh or dex rebuild
with an unchanged contract still changes the fingerprint; external audit 2026-07-13). The
environment comes from the meta skill's current.json."""
from __future__ import annotations

import hashlib
import json
from pathlib import Path

from . import __version__
from .paths import META_DATA, SKILLS_ROOT
from .worker import WorkerPool

# Must track frontend/protocol WEB_PROTOCOL_VERSION — bumped 1 -> 2 for the breaking
# ItemPanelEntryDto.key change, then 2 -> 3 for OppCacheDto.sets going REQUIRED -> optional so lean
# projections may omit it (a pre-3 SPA can't parse every v3 projection; external audit 2026-07-14),
# then 3 -> 4 because the default KO page requires the new lean `view=ko` / `oppko_<format>.json`
# resource before lazily loading full details, and 4 -> 5 for lazy per-Pokemon panel usage trends.
# The version is folded into the
# local deploymentId below so a DTO/mapper release is a new deployment identity even when skill data
# is unchanged.
WEB_PROTOCOL_VERSION = "5"

# What THIS bridge actually serves (llm.* are online-only; team.uep gates the local UEP
# session panel — sessions/artifacts/SSE plus its lazy chunk).
LOCAL_CAPABILITIES = [
    "dex.query", "meta.rank", "meta.detail", "meta.trend",
    "calc.damage", "calc.speedline",
    "team.validate", "team.matchup", "team.tune", "team.uep",
]

# Shipped data trees per fingerprint id (calc's "data" is the vendored engine+tables).
_DATA_TREES: dict[str, Path] = {
    "dex": SKILLS_ROOT / "pokemon-champions-dex/data",
    "meta": SKILLS_ROOT / "pokemon-champions-meta/data",
    "calc": SKILLS_ROOT / "ncp-damage-calculator/scripts/script_res",
    "team": SKILLS_ROOT / "pokemon-champions-team/data",
}

# calc's runtime is script_res (data tables) PLUS the two JS wrappers that drive them (buildPokemon /
# calculate / speedOne + the I/O contract). A wrapper BEHAVIOR change that leaves the `schema` output
# untouched must STILL change calc identity — otherwise a projection built on new wrapper logic shares a
# deploymentId with a SPA that ported the old wrapper into its worker (external audit 2026-07-14). Their
# content is folded into the calc fingerprint alongside script_res.
_CALC_WRAPPERS = [
    SKILLS_ROOT / "ncp-damage-calculator/scripts/ncp-calc-api.js",
    SKILLS_ROOT / "ncp-damage-calculator/scripts/ncp-speedline-api.js",
]


def _fingerprint(*docs: object) -> str:
    payload = json.dumps(docs, ensure_ascii=False, sort_keys=True).encode("utf-8")
    return hashlib.sha256(payload).hexdigest()[:16]


# SQLite writes transient sidecars (-wal / -shm, and -journal in rollback mode) NEXT TO the DB,
# inside the hashed data tree. A WAL-mode DB mutates them on mere READS (and checkpoint-on-close
# rewrites the -wal), so hashing them made the deployment identity DRIFT the moment any process
# queried the dex — projection built at identity A, first served query flipped it to B, and the
# atomic-binding guard then refused to render (external report 2026-07-14). They are not part of
# the shipped data VERSION (the .sqlite file is); exclude them so identity is stable across reads.
_TRANSIENT_SUFFIXES = ("-wal", "-shm", "-journal")


def _data_digest(tree: Path) -> str:
    """Content hash over a shipped data tree (relative path + bytes, sorted), excluding transient
    SQLite sidecars. ~40MB across all four skills, computed once per process on the first
    capabilities request."""
    h = hashlib.sha256()
    if tree.is_dir():
        # Path ordering follows the host flavour (WindowsPath compares case-insensitively while
        # PosixPath does not). The shipped trees contain mixed-case names such as damage_MASTER.js
        # and M-4_double.json, so relying on Path.__lt__ made identical bytes produce a different
        # deploymentId on the Windows projection builder and Linux VPS. Sort the wire-format key
        # explicitly: a POSIX relative-path string has the same order on every host.
        for p in sorted(tree.rglob("*"), key=lambda item: item.relative_to(tree).as_posix()):
            if p.is_file() and not p.name.endswith(_TRANSIENT_SUFFIXES):
                h.update(p.relative_to(tree).as_posix().encode("utf-8"))
                h.update(b"\0")
                h.update(p.read_bytes())
    return h.hexdigest()


def _files_digest(paths: list[Path]) -> str:
    """Content hash over an explicit file list (name + bytes, name-sorted) — for the calc wrappers,
    which live one level up from the hashed script_res tree."""
    h = hashlib.sha256()
    for p in sorted(paths, key=lambda x: x.name):
        h.update(p.name.encode("utf-8"))
        h.update(b"\0")
        h.update(p.read_bytes())
    return h.hexdigest()


def calc_engine_digest() -> str:
    """SHA-256 over the vendored calc engine sources (script_res/*.js + the two JS wrappers), basename-
    sorted and newline-normalized, each entry `name\\0content` joined by \\0. MUST byte-match the SPA's
    build-time __ENGINE_DIGEST__ (frontend/web/vite.config.ts calcEngineDigest), so the online runtime binds
    the engine it embedded in its worker to the projection it serves — a projection rebuilt on a newer
    engine than the SPA carries is then caught at the handshake, not silently mis-computed (external
    audit 2026-07-14). Distinct from the calc FINGERPRINT (which also folds in the schema + is truncated
    for the deployment id); this is the raw-source binding the browser can reproduce. Hashes RAW BYTES
    (not decoded text): some data files carry non-UTF-8 bytes and JS vs Python replace them differently
    on decode, so a text hash wouldn't agree — normalize CRLF->LF at the byte level, hash `name\\0bytes\\0`."""
    script_res = SKILLS_ROOT / "ncp-damage-calculator/scripts/script_res"
    entries = sorted(list(script_res.glob("*.js")) + _CALC_WRAPPERS, key=lambda p: p.name)
    h = hashlib.sha256()
    for p in entries:
        h.update(p.name.encode("utf-8"))
        h.update(b"\0")
        h.update(p.read_bytes().replace(b"\r\n", b"\n"))
        h.update(b"\0")
    return h.hexdigest()


def assemble(pool: WorkerPool, deployment_id: str | None = None,
             ui_build_id: str | None = None) -> dict:
    data = {k: _data_digest(v) for k, v in _DATA_TREES.items()}
    fingerprints = {
        "dex": _fingerprint(pool.request_json("dex", ["schema"]), data["dex"]),
        "meta": _fingerprint(pool.request_json("meta", ["schema"]), data["meta"]),
        "calc": _fingerprint(pool.request_json("calc", ["schema"]),
                             pool.request_json("speedline", ["schema"]), data["calc"],
                             _files_digest(_CALC_WRAPPERS)),
        "team": _fingerprint(pool.request_json("team", ["schema"]),
                             pool.request_json("team", ["vocab", "--format", "json"]),
                             data["team"]),
    }
    if deployment_id is None:
        # Local bridge: the "deployment" IS the installed skill snapshot — derive its id from the
        # combined fingerprints AND the wire-protocol version, so a breaking DTO/mapper change (which
        # bumps WEB_PROTOCOL_VERSION) is a new deployment identity even when skill data is unchanged
        # (external audit 2026-07-14: DTO version was previously absent from this fingerprint).
        deployment_id = "local-" + _fingerprint(fingerprints, WEB_PROTOCOL_VERSION)[:12]
    current = json.loads((META_DATA / "current.json").read_text(encoding="utf-8"))
    season = current["current"]["season"]
    env = {"season": season, "rule": current["current"]["rule"]}
    as_of = (current.get("seasons", {}).get(season, {}).get("updated_at") or "")[:10]
    if as_of:
        env["asOf"] = as_of
    return {
        "webProtocolVersion": WEB_PROTOCOL_VERSION,
        "deploymentId": deployment_id,
        # Real build id once the wheel embeds the SPA dist; until then the bridge version.
        "uiBuildId": ui_build_id or f"pcui-{__version__}",
        "skillFingerprints": fingerprints,
        "environment": env,
        "capabilities": LOCAL_CAPABILITIES,
    }

File: frontend/protocol/src/capabilities.ts
/**
 * Capabilities handshake (frontend/design.md §3): the bridge / online API assembles this once at
 * startup; the SPA fetches it before rendering. Three version fields serve distinct roles —
 * `webProtocolVersion` is THIS contract's version, `uiBuildId` fingerprints the SPA build,
 * `skillFingerprints` identify the four installed skills' data/code. Capabilities only drive
 * front-end visibility; the server whitelists every endpoint independently (§7.5).
 */
import { z } from "zod";

// Bumped 1 → 2: ItemPanelEntryDto.key went from absent to REQUIRED (a breaking DTO change). Without
// a bump, a projection built by an older mapper (no item keys) would share a deploymentId with a
// newer SPA and pass the startup handshake, then throw at MetaDetailDtoSchema.parse. The version gate
// now rejects such a projection up front (external audit 2026-07-14). Bump on any breaking DTO/mapper
// change, not just skill-data changes.
// Bumped 2 → 3: OppCacheDto.sets went REQUIRED → optional so lean projections can omit it. A pre-3
// SPA (sets required) can no longer parse every v3 projection; the version gate makes that breaking
// contract change explicit. The current public projection includes de-attributed processed sets.
// Bumped 3 → 4: the default KO page now requires the lean `oppko_<format>.json` projection/API view
// and lazy-loads the old full cache for detail. A v4 SPA paired with a v3 projection would 404.
// Bumped 4 → 5: meta detail popovers now require the lazy per-Pokemon usage-trend endpoint/projection.
export const WEB_PROTOCOL_VERSION = "5";

export const SKILL_IDS = ["dex", "meta", "calc", "team"] as const;
export type SkillId = (typeof SKILL_IDS)[number];

/** Capability ids the current SPA knows how to surface. The wire format allows unknown ids
 * (forward compatibility): a newer server may advertise features an older SPA simply ignores. */
export const KNOWN_CAPABILITIES = [
  "dex.query", "meta.rank", "meta.detail", "meta.trend",
  "calc.damage", "calc.speedline",
  "team.validate", "team.uep", "team.matchup", "team.tune",
  "llm.qa", "llm.builder",
] as const;
export type KnownCapability = (typeof KNOWN_CAPABILITIES)[number];

export const CapabilitiesSchema = z.object({
  webProtocolVersion: z.string().min(1),
  deploymentId: z.string().min(1),
  uiBuildId: z.string().min(1),
  skillFingerprints: z.record(z.enum(SKILL_IDS), z.string().min(1)),
  environment: z.object({
    season: z.string().min(1),
    rule: z.string().min(1),
    asOf: z.string().optional(),
  }),
  capabilities: z.array(z.string()),
});
export type Capabilities = z.infer<typeof CapabilitiesSchema>;

export type CompatibilityResult =
  | { ok: true; capabilities: Capabilities }
  | { ok: false; reasons: string[] };

/** Validate a handshake payload and gate on the protocol version. The SPA calls this before
 * trusting anything else the server says; a mismatch renders the incompatibility screen
 * instead of half-working pages. */
export function checkCompatibility(raw: unknown): CompatibilityResult {
  const parsed = CapabilitiesSchema.safeParse(raw);
  if (!parsed.success) {
    return {
      ok: false,
      reasons: parsed.error.issues.map((i) => `${i.path.join(".") || "$"}: ${i.message}`),
    };
  }
  if (parsed.data.webProtocolVersion !== WEB_PROTOCOL_VERSION) {
    return {
      ok: false,
      reasons: [
        `web protocol version ${parsed.data.webProtocolVersion} is not supported ` +
        `(this build speaks ${WEB_PROTOCOL_VERSION})`,
      ],
    };
  }
  return { ok: true, capabilities: parsed.data };
}

export const hasCapability = (c: Capabilities, id: KnownCapability): boolean =>
  c.capabilities.includes(id);
